### PR TITLE
:sparkles: Source Bing Ads: custom reports

### PR DESCRIPTION
**Pull request by darynaishchenko**
## What
Certification:
- custom reports for report-based connector
- backoff time <10min
- actionable error messages

## How
- Added custom reports generation (UI screenshots added below)
- decreased max_tries for backoff
- updated error message for invalid dev token case
image

image

image

image

### CODE REVIEW

File: airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json [hunk 1]
```diff
@@ -69,6 +69,64 @@
...
+              "minItems": 1
+            },
+            "report_aggregation": {
+              "title": "Aggregation",
+              "description": "A list of available aggregations.",
+              "type": "string",
+              "items": {
+                "title": "ValidEnums",
+                "description": "An enumeration of aggregations.",
+                "enum": ["Hourly", "Daily", "Weekly", "Monthly"]
```
Comment (artem1205) [inline airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json:118]: Enum should be extended:
https://learn.microsoft.com/en-us/advertising/reporting-service/reportaggregation?view=bingads-13

File: airbyte-integrations/connectors/source-bing-ads/source_bing_ads/source.py [hunk 2]
```diff
@@ -83,17 +86,66 @@ def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) ->
...
+        custom_reports = self.get_custom_reports(config, client)
+        for custom_report in custom_reports:
+            try:
+                for account in Accounts(client, config).read_records(SyncMode.full_refresh):
+                    list(
+                        custom_report.read_records(
+                            sync_mode=SyncMode.full_refresh,
+                            stream_slice={"account_id": account["Id"], "customer_id": account["ParentCustomerId"]},
+                        )
+                    )
```
Comment (artem1205) [inline airbyte-integrations/connectors/source-bing-ads/source_bing_ads/source.py:111]: Check connection will do full sync here?
We can use parsed `/bingads/v13/proxies/production/reporting_service.xml` to validate input data (and/or try to use `enum` values from there).
The only potential errors will left and should be handled during the read:
- combination of columns (see Column restrictions) 
- combination of columns with report_aggregation

**Review by artem1205 — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json [hunk 3]
```diff
@@ -69,6 +69,64 @@
...
+                "Account Performance",
+                "AdDynamicTextPerformanceReport",
+                "custom report"
+              ]
+            },
+            "reporting_object": {
+              "title": "Reporting Data Object",
+              "description": "The name of the the object derives from the ReportRequest object. You can find it in Bing Ads Api docs - Reporting API - Reporting Data Objects.",
+              "type": "string",
+              "examples": [
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json:96]: Is it possible to `enum` these data objects for the user so we could have the dropdown selection?

File: airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json [hunk 4]
```diff
@@ -69,6 +69,64 @@
...
+                "AccountPerformanceReport",
+                "AdDynamicTextPerformanceReport"
+              ]
+            },
+            "report_columns": {
+              "title": "Columns",
+              "description": "A list of available report object columns. You can find it in description of reporting object that you want to add to custom report.",
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json:103]: Please provide the root link where the columns can be seen and taken, so the customer can easily follow the link and check them out.

File: airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json (hunk 3 above)
Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json:96]: yes, all data objects are declared in [docs](https://learn.microsoft.com/en-us/advertising/reporting-service/reporting-data-objects?view=bingads-13), but it's more than 30, I'm not sure that is user-friendly to show such a big dropdown section

File: airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json (hunk 4 above)
Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json:103]: all reporting data objects have different columns, so the link relates to data object that was specified by user

Comment (bazarnov) [reply airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json:103]: Yes, unfortunately, this is the limitation for the dependent dropdowns currently, it's fine to leave as is, but we need to provide the link that could describe the columns available for each data object and aggregation type. It's better to mention this info in the public docs for the user.

File: airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json (hunk 3 above)
Comment (bazarnov) [reply airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json:96]: Let's add them. It's fine to have the dropdown and quickly find the relative object right off, then go to the docs and try to understand what the object should be.

File: airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json (hunk 1 above)
Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-bing-ads/source_bing_ads/spec.json]: extended with all, except Summary and Yearly

File: airbyte-integrations/connectors/source-bing-ads/source_bing_ads/source.py (hunk 2 above)
Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-bing-ads/source_bing_ads/source.py]: updated, now it  validates object name and columns, and combination of columns with report_aggregation validates while read

**Review by artem1205 — APPROVED** [review]

**Pull request merged**